Add packages and populate with module names

tlegen-k opened the pull request:
## Related Issues and Dependencies
https://github.com/thoth-station/datasets/pull/36 
…

## This introduces a breaking change

- [ ] Yes
- [x] No

## This should yield a new module release

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## This Pull Request implements

Add module names without namespaced calls from results to packages list to sync with database.

## Description

<!--- Describe your changes in detail -->

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fridex** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @fridex` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=thoth-station%2Fstorages).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/thoth-station/storages/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fridex"]} -->

tlegen-k commented:
/assign @tlegen-k

tlegen-k commented:
/label WIP

sesheta commented:
@tlegen-k: The label(s) `/label WIP` cannot be applied. These labels are supported: `community/discussion, community/group-programming, community/maintenance, community/question, deployment_name/ocp4-stage, deployment_name/ocp4-test, deployment_name/zero-prod, hacktoberfest, hacktoberfest-accepted, kind/cleanup, kind/demo, kind/deprecation, kind/documentation, kind/question, sig/advisor, sig/build, sig/cyborgs, sig/devops, sig/documentation, sig/indicators, sig/investigator, sig/knowledge-graph, sig/slo, sig/solvers, thoth/group-programming, thoth/human-intervention-required, thoth/potential-observation, tide/merge-method-merge, tide/merge-method-rebase, tide/merge-method-squash, triage/accepted, triage/duplicate, triage/needs-information, triage/not-reproducible, triage/unresolved`

<details>

In response to [this](https://github.com/thoth-station/storages/pull/2410#issuecomment-887855577):

>/label WIP


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pytest | d8373ec70271035f83180e094867fc045ba0575d | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/thoth-pytest-py38/1420137873813803008) | `/test thoth-pytest-py38`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

tlegen-k commented:
/label work-in-progress

sesheta commented:
@tlegen-k: The label(s) `/label work-in-progress` cannot be applied. These labels are supported: `community/discussion, community/group-programming, community/maintenance, community/question, deployment_name/ocp4-stage, deployment_name/ocp4-test, deployment_name/zero-prod, hacktoberfest, hacktoberfest-accepted, kind/cleanup, kind/demo, kind/deprecation, kind/documentation, kind/question, sig/advisor, sig/build, sig/cyborgs, sig/devops, sig/documentation, sig/indicators, sig/investigator, sig/knowledge-graph, sig/slo, sig/solvers, thoth/group-programming, thoth/human-intervention-required, thoth/potential-observation, tide/merge-method-merge, tide/merge-method-rebase, tide/merge-method-squash, triage/accepted, triage/duplicate, triage/needs-information, triage/not-reproducible, triage/unresolved`

<details>

In response to [this](https://github.com/thoth-station/storages/pull/2410#issuecomment-887858375):

>/label work-in-progress


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

tlegen-k commented:
/label do-not-merge/work-in-progress

sesheta commented:
@tlegen-k: The label(s) `/label do-not-merge/work-in-progress` cannot be applied. These labels are supported: `community/discussion, community/group-programming, community/maintenance, community/question, deployment_name/ocp4-stage, deployment_name/ocp4-test, deployment_name/zero-prod, hacktoberfest, hacktoberfest-accepted, kind/cleanup, kind/demo, kind/deprecation, kind/documentation, kind/question, sig/advisor, sig/build, sig/cyborgs, sig/devops, sig/documentation, sig/indicators, sig/investigator, sig/knowledge-graph, sig/slo, sig/solvers, thoth/group-programming, thoth/human-intervention-required, thoth/potential-observation, tide/merge-method-merge, tide/merge-method-rebase, tide/merge-method-squash, triage/accepted, triage/duplicate, triage/needs-information, triage/not-reproducible, triage/unresolved`

<details>

In response to [this](https://github.com/thoth-station/storages/pull/2410#issuecomment-887859328):

>/label do-not-merge/work-in-progress


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

sesheta commented:
@tlegen-k: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pytest | d8373ec70271035f83180e094867fc045ba0575d | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/thoth-pytest-py38/1420137873813803008) | `/test thoth-pytest-py38`
aicoe-ci/prow/pre-commit | d8373ec70271035f83180e094867fc045ba0575d | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1420137873406955520) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

tlegen-k commented:
Hi @pacospace 
Would you be able to take a look?
Next step will be to create new table linking PythonPackageVersion Table with the new table.

tlegen-k commented:
/label  community/group-programming

fridex commented on thoth/storages/graph/postgres.py:
```diff
@@ -5477,6 +5487,17 @@ def sync_solver_result(self, document: dict, *, force: bool = False) -> None:
+            import_package =[]
+            for package_call in document["result"]["tree"]["packages"]:
+                if package_call.count(".") == 0 && package_call[:4] != "test"
+                    && package_call[0] != "_":
```
```suggestion
                    and package_call[0] != "_":
```

fridex replied:
Could you give examples for entries that are skipped?

tlegen-k replied:
> Could you give examples for entries that are skipped?

Hi @fridex, thank you for helping!
Yes, when exploring [Thoth solver dataset v2 ](https://www.kaggle.com/thothstation/thoth-solver-dataset-v10?select=thoth-solver-dataset-v2.0) for report with documet_id = solver-rhel-8-py38-210713042150-50168537c03f93c5 tree:packages looks like:
```
['_plotly_future_',
 '_plotly_utils',
 'plotly',
 'plotlywidget',
 '_plotly_utils.colors',
 'plotly.colors',
 'plotly.data',
 'plotly.express',
 'plotly.figure_factory',
 'plotly.graph_objs',
 'plotly.io',
 'plotly.matplotlylib',
 'plotly.offline',
 'plotly.plotly',
 'plotly.validators',
 'plotly.graph_objs.area',
...
]
```

Here we would like to skip all hidden entry imports that start with '_'.
Also AFAIK it is the only one that has two modules, but `plotlywidget` is not on pypi.org.

fridex reviewed (commented):
Looking at this, I'm not sure if we do not break `get_or_create` logic and create redundant entries (one with packages and the other without).

pacospace commented on thoth/storages/graph/postgres.py:
```diff
@@ -4116,6 +4118,7 @@ def _create_python_package_version(
         python_version: Union[str, None],
         python_package_metadata_id: int = None,
         sync_only_entity: bool = False,
+        import_package: str,
```
mandatory parameter cannot be after optional one :) is it optional? then I would put `Optional[str] = None`

tlegen-k replied:
thank you)
I think it should be mandatory in this context 
corrected with c7996586173df63d59a6d46ac5a5063bc89bca20

pacospace commented on thoth/storages/graph/models.py:
```diff
@@ -72,7 +72,7 @@ class PythonPackageVersion(Base, BaseExtension):
     is_missing = Column(Boolean, nullable=False, default=False)
     provides_source_distro = Column(Boolean, nullable=False, default=True)
-
+    import_package = Column(Text, nullable=False)
```
nullable should be True, initially the database won't have these entries and we cannot assign a default in this case, wdyt?

tlegen-k replied:
> nullable should be True, initially the database won't have these entries and we cannot assign a default in this case, wdyt?

I agree, corrected in this commit d1a6c6a850b5e8f3e2577648f8c348225f492f8c.
Thanks for pointing!

tlegen-k commented:
Hi @pacospace and @fridex,

Regards possible number of modules in tree:packages, apparently there are more cases when there is more than 1 module.
As an example this [notebook](https://colab.research.google.com/drive/1jF84zOzVTd5ThW5kvSE7C8fz4M9xYuwf?usp=sharing).

To give part of nonunique imports:
```
[['plotly', 'plotlywidget'],
 ['chart_studio', 'plotly', 'plotlywidget'],
 ['Cython', 'pyximport'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['Cython', 'pyximport'],
 ['pkg_resources', 'setuptools'],
 ['Cython', 'pyximport'],
 ['plotly', 'plotlywidget'],
 ['pkg_resources', 'setuptools'],
 ['pkg_resources', 'setuptools'],
 ['caffe2', 'torch'],
...
```
They are under `nonunique_import_package` variable in the bottom of shared notebook.

tlegen-k commented:
gently ping @fridex. Would you be able to give a feedback?

fridex commented:
> gently ping @fridex. Would you be able to give a feedback?

As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).

tlegen-k commented:
Hi @fridex,
Here is another attempt in doing correctly.
I have created `ImportPackages` class to hold only `document_id` and `import_packages` and extended logic of `sync_solver_result` to correctly initialize instance of `ImportPackages`.

Are there any other data fields that might be beneficial to store in `ImportPackages`?

pacospace commented:
> > gently ping @fridex. Would you be able to give a feedback?
> 
> As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).

Should we make that mandatory? and resync all solvers with solver version that has that feature?

fridex commented on thoth/storages/graph/models.py:
```diff
@@ -1688,6 +1687,18 @@ class SIAggregated(Base, BaseExtension):
+
+    __tablename__ = "import_packages"
+
+    document_id = Column(Text, nullable=False)
```
This can be obtained from `PythonPackageVersion` directly so no need to keep it here.

```suggestion
```

fridex commented on thoth/storages/graph/models.py:
```diff
@@ -72,7 +72,6 @@ class PythonPackageVersion(Base, BaseExtension):
-
     # Relations
     dependencies = relationship("DependsOn", back_populates="version")
     solvers = relationship("Solved", back_populates="version")
```
You will need to add a relation to `ImportPackage`. I think you will need an association table for N-N relation here (as one package might provide the same import as another one):

https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many

tlegen-k replied:
resolved by 8a428279967604e0e603a2556370045264245cbd

fridex commented on thoth/storages/graph/postgres.py:
```diff
@@ -5477,6 +5477,34 @@ def sync_solver_result(self, document: dict, *, force: bool = False) -> None:
+
+            import_packages =[]
+            for package_call in document["result"]["tree"]["packages"]:
+                if package_call.count(".") == 0 and package_call[:4] != "test"
```
No need to have any filtering here on names. You can directly create `ImportPackage` per each entry in `document["result"]["tree"]["packages"]`.

tlegen-k replied:
resolved by 8a428279967604e0e603a2556370045264245cbd

fridex commented on thoth/storages/graph/models.py:
```diff
@@ -1688,6 +1687,18 @@ class SIAggregated(Base, BaseExtension):
     python_package_version_entity = relationship("PythonPackageVersionEntity", back_populates="si_aggregated_runs")
 
 
+class ImportPackages(Base, BaseExtension):
```
```suggestion
class ImportPackage(Base, BaseExtension):
```

fridex commented on thoth/storages/graph/models.py:
```diff
@@ -1688,6 +1687,18 @@ class SIAggregated(Base, BaseExtension):
+
+    document_id = Column(Text, nullable=False)
+
+    import_packages = Column(Text, nullable=False)
```
```suggestion
    import_package = Column(Text, nullable=False)
```

fridex commented:
> > > gently ping @fridex. Would you be able to give a feedback?
> > 
> > 
> > As stated above, not sure if this would be a good solution. `get_or_create` logic on models creates a new entry in tables if there is no matching entry with all the attributes. Your code extends `PythonPackageVersion` which is used in other places of the project - see the adjusted part for sync_adviser_result as an example. This would mean that we eventually lose linkage to entries in the database if `import_package` will be set in some cases and in some cases it will be set to None (duplicate entries might be created).
> 
> Should we make that mandatory? and resync all solvers with solver version that has that feature, so we don't create double entries? @fridex @tlegen-k

Making the field mandatory will not work - this information is present only in solver results and other sync methods also create `PythonPackageVersion` entries that would require a linkage to the newly introduced table.

pacospace commented on thoth/storages/graph/postgres.py:
```diff
@@ -5477,6 +5477,27 @@ def sync_solver_result(self, document: dict, *, force: bool = False) -> None:
                             )
 
+
+            # Sync imoprted package from result:tree
```
this should run below https://github.com/thoth-station/storages/blob/e6fbbc34f782b8555729cf9f8fe270b2753f341f/thoth/storages/graph/postgres.py#L5332 :)

tlegen-k replied:
resolved by 8e65f38da030758bc68904136a0c0888e7ed54aa

pacospace commented on thoth/storages/graph/models.py:
```diff
@@ -51,6 +51,12 @@
 )
 
 
+association_table = Table('association', Base.metadata,
```
let's follow some example we have in the databse like: https://github.com/thoth-station/storages/blob/e6fbbc34f782b8555729cf9f8fe270b2753f341f/thoth/storages/graph/models.py#L355

pacospace reviewed (commented):
Once changes are done, new alembic migration to be added to the PR :)

pacospace commented on thoth/storages/graph/postgres.py:
```diff
@@ -315,7 +316,7 @@ def connect(self):
             if not self.is_schema_up2date():
                 _LOGGER.debug("Database adapter connected, database is initialized")
-        except DatabaseNotInitialized as exc:
+        except Exception as exc:
```
why this has changed?

tlegen-k replied:
To be able to run locally on macOS. Will make as it was before sending to prod.

pacospace commented on thoth/storages/graph/models.py:
```diff
@@ -50,6 +50,12 @@
     EnvironmentTypeEnum.RUNTIME.value, EnvironmentTypeEnum.BUILDTIME.value, name="environment_type", create_type=True
 )
 
+association_table = Table(
```
this should be a class as all tables

tlegen-k replied:
logic of connection between `ImportPackage` and `PPV` was changed to Association Object as requested by using `FoundImportPackage` in commit b4147fd5152048337263037bd70f41c6a96026af

tlegen-k commented:
Hi @pacospace, thanks for continuous input!
I rewrote the model using association object instead of association table in b4147fd5152048337263037bd70f41c6a96026af.
Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
Next will be:
- [ ] Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
- [ ] Create final alembic migration
Let me know what you think.

fridex commented:
> Hi @pacospace, thanks for continuous input!
> I rewrote the model using association object instead of association table in [b4147fd](https://github.com/thoth-station/storages/commit/b4147fd5152048337263037bd70f41c6a96026af).
> Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
> Next will be:
> 
> * [ ]  Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
> * [ ]  Create final alembic migration
>   Let me know what you think.

Might be a good idea to also include an index on `import_package_name` given the expected size of the newly created table. 👍🏻

tlegen-k commented:
> > Hi @pacospace, thanks for continuous input!
> > I rewrote the model using association object instead of association table in [b4147fd](https://github.com/thoth-station/storages/commit/b4147fd5152048337263037bd70f41c6a96026af).
> > Also created function `get_python_package_version_import_packages_all` that returns `ImportPackage` rows for the given `import_name`.
> > Next will be:
> > 
> > * [ ]   Extend function so that: it outputs the PPV entries that are connected with the filtered 'ImportPackage`
> > * [ ]   Create final alembic migration
> >   Let me know what you think.
> 
> Might be a good idea to also include an index on `import_package_name` given the expected size of the newly created table. 👍🏻

Hi @pacospace and @fridex 
Added indexing in  739304e46a47c62c3c1e82c32383e086861525f8.
Final versions of `models.py` and `postgres.py`. Tested everything locally. Gives correct output.
For command `PYTHONPATH=. pipenv run python3 test_query.py` with following `test_query.py`:
```
from thoth.storages import GraphDatabase
from thoth.storages.sync import sync_solver_documents

# query part
graph = GraphDatabase()
graph.connect()


print(graph.get_python_package_version_import_packages_all("selinon"))
print(graph.get_python_package_version_import_packages_all("sklearn"))
print(graph.get_python_package_version_import_packages_all("dotenv"))
print(graph.get_python_package_version_import_packages_all("IPython"))
```
Output is: 
```
['selinon']
['scikit-learn']
['python-dotenv']
['ipython']
```
Which correctly identifies package name required to be installed.

sesheta commented:
@tlegen-k: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | d8373ec70271035f83180e094867fc045ba0575d | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1420137873406955520) | `/test pre-commit`
aicoe-ci/prow/pytest | 739304e46a47c62c3c1e82c32383e086861525f8 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/thoth-pytest-py38/1425892348092485632) | `/test thoth-pytest-py38`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sesheta commented:
@tlegen-k: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pytest | 739304e46a47c62c3c1e82c32383e086861525f8 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/thoth-pytest-py38/1425892348092485632) | `/test thoth-pytest-py38`
aicoe-ci/prow/pre-commit | 739304e46a47c62c3c1e82c32383e086861525f8 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1425892347790495744) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pacospace commented on thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py:
```diff
@@ -0,0 +1,54 @@
+    sa.ForeignKeyConstraint(['python_package_version_id'], ['python_package_version.id'], ondelete='CASCADE'),
+    sa.PrimaryKeyConstraint('python_package_version_id', 'import_package_id')
+    )
+    op.alter_column('has_symbol', 'software_environment_id',
```
please remove line 33 to 40 from the migration manually as they are not related to this. cc @KPostOffice that encountered similar problem.

tlegen-k replied:
removed in 152184ab2cafbc52b8c8171b91a5f8317e02d60e

pacospace commented on thoth/storages/graph/postgres.py:
```diff
@@ -6765,3 +6790,15 @@ def get_python_package_version_all(self, analysis_document_id: str) -> List[Dict
 
             return [{"package_name": i[1], "package_version": i[2], "location": i[0]} for i in query.all()]
+
+    def get_python_package_version_import_packages_all(self, import_name: str) -> str:
```
return type should be `List[str]`

tlegen-k replied:
fixed in 3f36204cb14d6f6cd76d9bdb7357c33dcfa5ad51

pacospace reviewed (commented):
Few comments and please put back parts that you commented/modified for testing. Very nice work 👏🏼

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | 3f36204cb14d6f6cd76d9bdb7357c33dcfa5ad51 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1425901880076144640) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pacospace commented on thoth/storages/data/alembic/versions/c1839dfcdf13_add_importpackage_foundimportpackage_.py:
```diff
@@ -0,0 +1,47 @@
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_constraint(None, 'python_package_version_entity_rule', type_='unique')
+    op.alter_column('has_symbol', 'versioned_symbol_id',
```
Also line 39 to 44 should be manually removed

tlegen-k replied:
was removed

tlegen-k commented:
> Few comments and please put back parts that you commented/modified for testing. Very nice work 👏🏼

Thank you for review @pacospace!
I made sure to return part that were commented to run locally.
What should I work on next?

pacospace commented on thoth/storages/graph/models.py:
```diff
@@ -1688,6 +1687,29 @@ class SIAggregated(Base, BaseExtension):
+    import_package_name = Column(Text, nullable=True)
+
+    python_package_versions = relationship("FoundImportPackage", back_populates="import_package")
+
```
I actually don't see the index created for `import_package_name` as @fridex suggested. It's missing the migration also indeed, there should be something like `op.create_index`

tlegen-k replied:
> I actually don't see the index created for `import_package_name` as @fridex suggested. It's missing the migration also indeed, there should be something like `op.create_index`

I understood as `id` field is index in line 1695:
`id = Column(Integer, primary_key=True, autoincrement=True, nullable=False)`

![Screen Shot 2021-08-12 at 5 33 48 PM](https://user-images.githubusercontent.com/58737562/129273057-9b097aa5-fc07-4400-bc9f-2ae253efc6fc.png)

Or if it is not that i can definitely work on that.

pacospace replied:
See for example: https://github.com/thoth-station/storages/blob/61b9de2fa41a31547bb1b45ce7b21276455e70e7/thoth/storages/graph/models.py#L215. I think @fridex meant to have one specific for that, the ID has an index by default I think.

tlegen-k replied:
Now clear. Implemented in d8d35af04d26d99a2f8187c78f267a6639e026a6
And created new alembic migration with adjustments to it that you pointed out previously

tlegen-k replied:
> See for example:
> 
> https://github.com/thoth-station/storages/blob/61b9de2fa41a31547bb1b45ce7b21276455e70e7/thoth/storages/graph/models.py#L215
> . I think @fridex meant to have one specific for that, the ID has an index by default I think.

Now clear. Implemented in d8d35af04d26d99a2f8187c78f267a6639e026a6
And created new alembic migration with adjustments to it that you pointed out previously
Let me know if there is anything else)

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | 1af60afede2225a116c04c98983fd9284cc5d522 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1425929750001487872) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

goern commented:
@all what is the status of this one? does it need a little love before ready to merge?

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | d8d35af04d26d99a2f8187c78f267a6639e026a6 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1426174490953388032) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

fridex commented on thoth/storages/graph/postgres.py:
```diff
@@ -5477,6 +5501,7 @@ def sync_solver_result(self, document: dict, *, force: bool = False) -> None:
                                 marker_evaluation_result=dependency.get("marker_evaluation_result", True),
                             )
 
+
```
```suggestion
```

fridex commented on thoth/storages/graph/postgres.py:
```diff
@@ -5440,6 +5442,28 @@ def sync_solver_result(self, document: dict, *, force: bool = False) -> None:
+                        import_package_id=import_package.id,
+                    )
+
+
```
```suggestion
```

fridex reviewed (commented):
Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.

Nice work! 💯 👍🏻

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | 3bea90a5e6b9fb5e4078591e5fd4ba330689288b | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1427397854091022336) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

tlegen-k commented:
> Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> 
> Nice work! 💯 👍🏻

Thank you @fridex! Have squashed 26 commits using git rebase.

sesheta commented:
@tlegen-k: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
aicoe-ci/prow/pre-commit | bbd9257d20859d1eec2f427e7ea6abffd75d66a8 | [link](https://prow.operate-first.cloud/view/s3/ci-prow-artifacts--e2081a4d-fd23-4347-93d5-f33e0604d95d/prow-logs/pr-logs/pull/thoth-station_storages/2410/pre-commit/1427413154073153536) | `/test pre-commit`

[Full PR test history](https://prow.operate-first.cloud/pr-history?org=thoth-station&repo=storages&pr=2410). [Your PR dashboard](https://prow.operate-first.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Atlegen-k). Please help us and [open an issue](https://github.com/thoth-station/storages/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pacospace commented:
> > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > Nice work! 💯 👍🏻
> 
> Thank you @fridex! Have squashed 26 commits using git rebase.

Can you please adjust to make pre-commit happy? :)

tlegen-k commented:
> > > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > > Nice work! 💯 👍🏻
> > 
> > 
> > Thank you @fridex! Have squashed 26 commits using git rebase.
> 
> Can you please adjust to make pre-commit happy? :)

Totally overlooked that one)
Thanks for pointing out @pacospace 
Fixed now.  Only needs approval labels.

pacospace commented:
> > > > Nice! The code looks good to me (haven't tried it thought). Could you please squash commits to keep the git history clean? It might help in the future with debugging.
> > > > Nice work! 💯 👍🏻
> > > 
> > > 
> > > Thank you @fridex! Have squashed 26 commits using git rebase.
> > 
> > 
> > Can you please adjust to make pre-commit happy? :)
> 
> Totally overlooked that one)
> Thanks for pointing out @pacospace
> Fixed now. Only needs approval labels.

Thanks!! We just miss rebase after https://github.com/thoth-station/storages/pull/2384#issuecomment-901445109 was merged and please adjust the alembic migration number now, because that PR already introduced a migration :)

fridex commented on thoth/storages/graph/postgres.py:
```diff
@@ -6786,3 +6809,15 @@ def get_python_package_version_all(self, analysis_document_id: str) -> List[Dict
+                .filter(ImportPackage.import_package_name == import_name)
+                .filter(PythonPackageVersion.id == FoundImportPackage.python_package_version_id)
+            )
+            return [i.package_name for i in query.all()]
```
Can the query return `package_name`, `package_version`, `index_url` triplet so the package is uniquely identified?

fridex reviewed (approved):
lgtm, thanks @tlegen-k 💯 👍🏻

sesheta commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/thoth-station/storages/pull/2410#pullrequestreview-741306786" title="Approved">fridex</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=thoth-station%2Fstorages).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/thoth-station/storages/blob/master/OWNERS)~~ [fridex]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

tlegen-k commented:
> lgtm, thanks @tlegen-k 💯 👍🏻

Thank you for helping!
